Changed MySql New-AzMySqlFlexibleServer cmdlet and Added Get-AzMySqlFlexibleServerLocationBasedCapability

DaeunYim opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

<!-- Please add a brief description of the changes made in this PR -->

## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

msJinLei commented on src/MySql/test/Remove-AzMySqlFlexibleServer.Tests.ps1:
```diff
@@ -14,9 +14,11 @@ while(-not $mockingPath) {
-            #[SuppressMessage("Microsoft.Security", "CS002:SecretInNextLine")]
-            $password = 'Pasword01!!2020' | ConvertTo-SecureString -AsPlainText -Force
-            New-AzMySqlFlexibleServer -ResourceGroupName $env.resourceGroup -Name $env.serverName -AdministratorUserName mysql_test -AdministratorLoginPassword $password 
+            If ($TestMode -eq 'live' -or $TestMode -eq 'record') {
```
You don't need to add the condition here.

msJinLei commented on src/MySql/custom/New-AzMySqlFlexibleServer.ps1:
```diff
@@ -147,60 +183,418 @@ function New-AzMySqlFlexibleServer {
+    }
+    
+    #Handle Vnet, Subnet scenario
+    If ($NetworkParameters.ContainsKey('Vnet') -Or $NetworkParameters.ContainsKey('Subnet')) {
```
Please keep case consistent in the module

msJinLei commented on src/MySql/custom/New-AzMySqlFlexibleServer.ps1:
```diff
@@ -147,60 +183,418 @@ function New-AzMySqlFlexibleServer {
+            ElseIf (![string]::IsNullOrEmpty($FirewallRuleName)) {
+                $Msg += "`"firewallName`": `"$FirewallRuleName`""
+            }
+            Write-Host $Msg
```
Please do not print the server information. PowerShell will print output automatically if no piping.

DaeunYim replied:
I want to also print out the datatabase name, firewall name, password that will not be available from the server output.

DaeunYim replied:
Answered together with the below comment

msJinLei commented on src/MySql/custom/New-AzMySqlFlexibleServer.ps1:
```diff
@@ -147,60 +183,418 @@ function New-AzMySqlFlexibleServer {
+            }
+            Write-Host $Msg
+
+            return $Server
```
Add all the field you want to print in the directlve.

DaeunYim replied:
New-AzMySqlFkexibleServer: I want to show  the password, firewall rule name, database name that are not available in the server object. 
By this: "Add all the field you want to print in the directlve to type server", Do you mean that I can add this to the returned object? Or is there any way I can add that??

msJinLei replied:
You can create a PSCustomObject which includes all the properties you want and return it.

msJinLei replied:
As we disscused, you can create a new class implements IServerAutoGenerated using c#, add the customerized property you want and use it in New-AzMySqlFlexibleServer. Please refer to the custom folder of CloudService

msJinLei replied:
@DaeunYim  check with the codes, find that MySqlIdentity is not the base class of IServerAutoGenerated. It can be converted from some specified types.

```csharp
            if (null == sourceValue)
            {
                return true;
            }
            global::System.Type type = sourceValue.GetType();
            // we allow string conversion too.
            if (type == typeof(global::System.String))
            {
                return true;
            }
            if (typeof(global::System.Management.Automation.PSObject).IsAssignableFrom(type))
            {
                // we say yest to PSObjects
                return true;
            }
            if (typeof(global::System.Collections.IDictionary).IsAssignableFrom(type))
            {
                // we say yest to Hashtables/dictionaries
                return true;
            }
```
You could try to create a psobject rather than pscustomobject

msJinLei commented on src/MySql/custom/New-AzMySqlFlexibleServer.ps1:
```diff
@@ -147,60 +183,418 @@ function New-AzMySqlFlexibleServer {
+                $PSBoundParameters.ResourceGroupName = Get-RandomNumbers -Prefix 'group' -Length 10
+                $Msg = "Creating Resource Group {0}..." -f $PSBoundParameters.ResourceGroupName
+                Write-Host $Msg
+                $null = New-AzResourceGroup -Name $PSBoundParameters.ResourceGroupName -Location $PSBoundParameters.Location
```
Use should process before create a resource group and use $Force parameter to skip the should process judgement
https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.1

DaeunYim replied:
Added!

msJinLei commented on src/MySql/custom/New-AzMySqlFlexibleServer.ps1:
```diff
@@ -147,60 +183,418 @@ function New-AzMySqlFlexibleServer {
+            $PSBoundParameters.AdministratorLoginPassword = . "$PSScriptRoot/../utils/Unprotect-SecureString.ps1" $PSBoundParameters['AdministratorLoginPassword']
+
+            If (!(Get-Module -ListAvailable -Name Az.Resources)) { Throw 'Please install Az.Resources module by entering "Install-Module -Name Az.Resources"'}
+            Else { Import-Module -Name Az.Resources }
```
You can Import-Module -Name Az.Resources directly. A error will ocurr if the module is not installed

DaeunYim replied:
Will do!

msJinLei commented on src/MySql/custom/New-AzMySqlFlexibleServer.ps1:
```diff
@@ -147,60 +183,418 @@ function New-AzMySqlFlexibleServer {
+    Catch {
+        $Msg = "Creating new vnet {0} in resource group {1}" -f $Parameters.VnetName, $Parameters.ResourceGroupName
+        Write-Host $Msg
+        New-AzVirtualNetwork -Name $Parameters.VnetName -ResourceGroupName $Parameters.ResourceGroupName -Location $Parameters.Location -AddressPrefix $Parameters.VnetPrefix -Force
```
Support should process for create virtual network

DaeunYim replied:
Added!

msJinLei reviewed (changes requested):
Please restore examples/New-AzMySqlVirtualNetworkRule.md to the revision before flexible server? The file is incorrectly modified recently.

msJinLei commented:
@DaeunYim Please use @autorest/powershell@3.0.409 to generate codes. Some tests may fail and you need to re-record them.

DaeunYim commented on src/MySql/examples/New-AzMySqlVirtualNetworkRule.md:
````diff
@@ -1,11 +1,18 @@
-### Example 1: Create a new MySql server Virtual Network Rule
+### Example 1: {{ Add title here }}
 ```powershell
````
Will do!

DaeunYim commented on src/MySql/custom/Get-AzMySqlFlexibleServerLocationBasedCapability.ps1:
```diff
@@ -0,0 +1,122 @@
+                    $Keys = $Skus.SupportedServerVersion[0].SupportedVcore
+                    
+                    ForEach ($Key in $Keys) {
+                        $NewEntry = @{}
```
Will do!

DaeunYim commented on src/MySql/custom/Get-AzMySqlFlexibleServerLocationBasedCapability.ps1:
```diff
@@ -0,0 +1,122 @@
+                    Throw "No SKU info for this location"
+                }
+            }
+            $TableResult | ForEach-Object {[PSCustomObject]$_} | Format-Table 'SKU', 'Tier', 'Memory', 'vCore'  -AutoSize
```
Thanks!!

DaeunYim commented on src/MySql/docs/New-AzMySqlVirtualNetworkRule.md:
```diff
@@ -23,17 +23,23 @@ Creates or updates an existing virtual network rule.
 
 ## EXAMPLES
 
-### Example 1: Create a new MySql server Virtual Network Rule
```
Will do

DaeunYim commented on src/MySql/custom/Get-AzMySqlFlexibleServerConnectionString.ps1:
```diff
@@ -0,0 +1,139 @@
+        [System.Management.Automation.SwitchParameter]
+        # Use the default credentials for the proxy.
+        ${ProxyUseDefaultCredentials}
+    )
```
Will do

DaeunYim commented:
> Please restore examples/New-AzMySqlVirtualNetworkRule.md to the revision before flexible server? The file is incorrectly modified recently.

Restored!

> @DaeunYim Please use @autorest/powershell@3.0.409 to generate codes. Some tests may fail and you need to re-record them.

Rebuilt...!

msJinLei commented:
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/791

msJinLei commented on src/MySql/custom/New-AzMySqlFlexibleServerFirewallRule.ps1:
```diff
@@ -19,124 +19,125 @@ Creates a new firewall rule for MySQL flexible server
-    [System.Management.Automation.SwitchParameter]
-    # Use the default credentials for the proxy
-    ${ProxyUseDefaultCredentials}
+    [OutputType([Microsoft.Azure.PowerShell.Cmdlets.MySql.Models.Api20171201.IFirewallRule])]
```
@DaeunYim Please revert the change of this file

msJinLei commented:
> > Please restore examples/New-AzMySqlVirtualNetworkRule.md to the revision before flexible server? The file is incorrectly modified recently.
> 
> Restored!
> 
> > @DaeunYim Please use @autorest/powershell@3.0.409 to generate codes. Some tests may fail and you need to re-record them.
> 
> Rebuilt...!
@DaeunYim 
- Please revert the New-AzMySqlFlexibleServerFirewallRule.ps1
- Please use pscustomobject to include all the property you want for New-AzMySqlFlexiblerServer
Please submit the revision as soon as possible since today is the code freeze day.

msJinLei commented:
@DaeunYim can you summarize the changes into some items as release notes of this feature?

DaeunYim commented:
> @DaeunYim
> 
> * Could you check the failed the [test cases](https://dev.azure.com/azure-sdk/public/_build/results?buildId=662521&view=logs&j=b17395f6-68a3-5682-0476-d3f6f1043109&t=e50c9eb1-7f19-5aa6-670b-cc46e51b352e) ? are you sure you are using powershell@3.0.409? `autorest.cmd --use:@autorest/powershell@3.0.409`
> * can you summarize the changes into some items as release notes of this feature?

Yes it did

DaeunYim commented:
> * can you summarize the changes into some items as release notes of this feature?


Az.MySql 
Updated New-AzMySqlFlexibleServer cmdlet to provide easy resource group and network management within RDBMS flexible server management. 
Added new cmdlets Get-AzMySqlFlexibleServerLocationBasedCapability and Get-AzMySqlConnectionString

msJinLei reviewed (changes requested):
@DaeunYim The doc/Get-AzMySqlFlexibleServerConnectionString.md doens't updated

msJinLei commented on src/MySql/custom/New-AzMySqlFlexibleServer.ps1:
```diff
@@ -19,188 +19,589 @@ Creates a new MySQL flexible server
+`"databaseName`": `"$DEFAULT_DB_NAME`",
+`"id`": `"/subscriptions/$($PSBoundParameters.SubscriptionId)/resourceGroups/$($PSBoundParameters.ResourceGroupName)/providers/Microsoft.DBForMySql/flexibleServers/$($PSBoundParameters.Name)`",
+`"location`": `"$($PSBoundParameters.Location)`",
+`"password`": `"$($PSBoundParameters.AdministratorLoginPassword)`",
```
For security issue, password cannot be printed as plain text. You should use security string.

DaeunYim replied:
The password is autogenerated that we need to inform the users what password is. What is the best way to show it to the customers?
